docs/prover/approx/hashing.md

```diff
@@ -0,0 +1,182 @@
+Instead, the Keccak hash function is modeled as an arbitrary function that is _injective with large gaps_. 
+
+The hash function `hash` being injective with large gaps means that on distinct inputs `x` and `y`
+  - the hashes `hash(x)` and `hash(y)` are also distinct, and
+  - the gap between `hash(x)` and `hash(y)` is large enough that every additive term `hash(x) + o` that occurs in the program is also distinct from `hash(y)`.
+
+## Background: The Solidity Storage Model
+
+For instance consider this contract:
+```solidity
+contract C {
+  uint i;                  // slot 0
+  uint[] a;                // slot 1
+  mapping(uint => uint) m; // slot 2
+  // ...
+  function foo() {
+	// ...
+	i = u;    // sstore(0, u)
+    // ... 
+    a[j] = w; // sstore(hash(1) + j, w)
+    // ... 
+    m[k] = v; // sstore(hash(2, k), v)
+    // ... 
+  }
+}
+```
+
+Here we can see how storage is laid out by solidity.
+The occurrences of `sstore(x, y)` in the line comments above denote a storage update of storage address `x` to value `y`.
+The scalar `i` is stored at storage address `0`, which is derived from it's slot number in the contract (slots are numbered in order of appearance in the source code).
+The array `a` is stored contiguously, starting from slot `hash(1)`.
+The entries of mapping `m` are spread out over storage; their locations are computed as the hash of the mapping's storage slot and the key at which the mapping is being accessed; thus the storage slot used for the entry of `m` under key `k` is computed as `hash(2, k)`.
+
+We can see that non-collision of hashes is essential for storage integrity. E.g., if `hash(1) + j` was equal to `hash(2, k)` then the operations on `a` an `m` would interfere with each other.
+
+Also note that the initial storage slots are reserved, i.e., we make sure that no hash value ends up colliding with slots 0 to 10000.
+
+
+## Modeling the Keccak Function (bounded case)
+
+The Certora Prover models the Keccak hash function as an arbitrary function that is _injective with large gaps_.
+That means that if `x != y` then `hash(x) != hash(y)`, but also that for all additive offsets `o` that actually occur in the program `hash(x) + o != hash(y)`.
+
+These constraints are enough for the solidity storage model to work as expected. However when hashes are compared, they might show different behavior from the actual Keccak function (e.g. `hash(x) > hash(y)` is always true or false for given `x` and `y`, but in our formulas, the SMT solver is free to choose a function for `hash` that makes the formula satisfiable.). 
+We have not observed a practical use case yet where the numeric values of the hash function play a role, thus we chose this modeling for tractability reasons.
+
+
+### Example (Imprecision of Modeling)
+
```

Comment:
Might also be worth including a super-simple example, like
```
// sol
function getHash() ... { return keccak("hello"); }

rule checkHash {
  assert getHash() == 0x....;
}
```

(filled in with actual syntax and examples)

Reply:
just to be sure: 
what should that example demonstrate? also the imprecision? i.e. that the actual `keccak(constant)` is some value and we get a violation when checking that? --> makes sense ...

Reply:
added an example that shows that we can neither prove nor disprove that the hash of a fixed string has the actual keccak value